SethBurns/bare-minimum
# Feature/new task view

Issue number:
This PR closes issue #3 

Feature added:
Full functionality of the task-selection-view:
- When user clicks a category, they're taken to the page where they see a randomly picked task from that category.
- User can then either refresh the choice or save that task.
- If user refreshes, they won't see the same task until all the tasks in that category have been viewed at least once.
- If user saves the task, and the save was successful, user will see appropriate feedback display for 1 second and then see another randomly picked task.
- If the save was unsuccessful, user will see appropriate error handling. 
- User has buttons to view the homepage and the tasklist.
- The elements are responsive to screensize

How did you implement the features:
- The code might be a lot because there's a lot of functionality but essentially it's just cohesive use of states and useEffect with the right dependencies.
- When we initially fetch the tasks per category, we store them in currentTasks state. This array is an array of objects with a property of "seen". Initially this is set to false.
- Another state of unseenTasks is set (using useEffect) everytime "currentTasks" changes. It filters the currentTasks by only selecting the tasks that have `seen===false`
- Then, again using useEffect (dependent on unseenTasks), we set a `currentTask` (singular) which is a random pick from `unseenTasks`. 
- We use the state `tasksToShow` to conditionally render any task. If `tasksToShow` is false, then error message is displayed saying that user has saved all tasks.
- `displaySavedResponse` is a boolean which, if true, allows conditional animation of feedback after save button is clicked. We used setTimeout for this to toggle it back to false (because feedback has to disappear after 1 second).
- `saveSuccessful` is another boolean state. `saveResponse` is a string state denoting the actual feedback string we want to show. Depending on the boolean value of `saveSuccessful` we set the string of `saveResponse`.
- Finally, we have `saveIcon` which can be assigned three different icons of different colors to visually assert the save feedback (pre-click: purple, post-click success: green, post-click failure: red)
- The key to absorb this code would be to follow the dependencies of all useEffects :) 

UI screenshots:

![Screenshot 2023-07-22 at 11 26 21 PM](https://github.com/SethBurns/bare-minimum/assets/122247155/bcda999a-19cd-4e52-a577-7e7cfccbd285)
![Screenshot 2023-07-22 at 11 26 08 PM](https://github.com/SethBurns/bare-minimum/assets/122247155/f84fd030-4737-44f7-b334-778012494904)


Installation notes:
just run `npm start`

Testing notes:
- Click on a category and browse through the tasks at a slow pace. 
- When you hover on refresh icon, the icon should rotate. 
- When you click on refresh, a new task should be displayed.
- Keep a track of the tasks you're seeing. You should not see a repeat until you've seen all tasks atleast once.
- Now start saving the tasks, at a slow pace. Observe the feedback displayed via text(saved) and the icon color (green).
- Once all tasks are saved, you should be able to see a message indicating that all tasks are saved. And the buttons to accept/deny should disappear.
- Refresh the page, now try to spam click the save button. If you click the save icon twice within 1 second, you should see error handling. Check the tasklist to ensure that the task was saved just once.
- Resize the screen. Ensure that everything works and still looks visually aesthetic. Please note that the icon hover effects will be removed via media queries because hover can not happen on phones.

Any other notes:
